### PR TITLE
[WIP] Feature/password recovery

### PR DESCRIPTION
**Pull request by lkwdwrd**
This will not be merged until the features the PR relys on are finished.

This is getting the initial code in place for supporting brokered access to credentials in the identity service. This feature set initially is in place to support password resets, but is also the cornerstone that alternate login methods will likely be built on (OIDC, etc...).

- Notes are updated to support EACP definitions in options
- On registration, extra notes for the identity broker are written
- New methods for initializing the broker flow, and completing the broker flow
- Aliases to the broker methods are added to make the password reset use-case more clear

Before merge, this needs to be hooked up with the actual features, and tested.

### CODE REVIEW

File: lib/identity/index.js [hunk 1]
```diff
@@ -89,7 +89,7 @@ export default class Identity extends CryptoConsumer {
     return this.constructor.StorageClient
   }
 
-  async register(username, password, token) {
+  async register(username, password, token, email) {
```
Comment (eddier) [inline lib/identity/index.js]: Is email a required param?  I would think most times the username will be an email address so wondering if it's a good idea to consolidate this?

Comment (lkwdwrd) [reply lib/identity/index.js]: doing it as two give additional flexibility -- the exact signature can change. I agree that most of the time email and username will be the same, but in the event they aren't, they need to be managed separately. I can even have the method check if email is defined and copy over username to email, meaning if the username is email, the parameter does not need to get sent.

File: lib/identity/index.js [hunk 2]
```diff
@@ -145,17 +145,65 @@ export default class Identity extends CryptoConsumer {
       }
     )
     /* eslint-enable */
-    /* eslint-disable-next-line no-warning-comments */
-    // TODO: Write recovery note based on final parameters
+    const brokerClientID = '00000000-0000-0000-0000-000000000000'
```
Comment (willmichael) [inline lib/identity/index.js]: Do we want this to broker flow to be optional//higher security tradeoffs?

File: lib/utils/credentials.js [hunk 3]
```diff
@@ -1,7 +1,16 @@
...
-  const noteID = await crypto.genericHash(username + idConfig.realmId)
+export async function deriveNoteCreds(
+  idConfig,
+  crypto,
+  username,
+  password,
+  broker = false
+) {
+  const nameSeed = broker
+    ? `broker${username}${idConfig.realmId}`
```
Comment (willmichael) [inline lib/utils/credentials.js]: This was the name above `broker${username}${this.config.realmId}key` but now its missing `key` did I miss a note?

File: lib/types/emailEACP.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * @copyright Copyright (c) 2018-19 Tozny, LLC (https://tozny.com)
+ */
+
+'use strict'
+
+import Serializable from './serializable'
+
+/**
+ * NoteOptions represents optional values that are not required for creating a note,
```
Comment (willmichael) [inline lib/types/emailEACP.js]: copy pasta comment?

File: lib/types/noteAccessEACP.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ * @copyright Copyright (c) 2018-19 Tozny, LLC (https://tozny.com)
+ */
+
+'use strict'
+
+import Serializable from './serializable'
+
+/**
+ * NoteOptions represents optional values that are not required for creating a note,
```
Comment (willmichael) [inline lib/types/noteAccessEACP.js]: needs relevant comment

File: lib/types/eacp.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+ */
+
+'use strict'
+
+import Serializable from './serializable'
+import EmailEACP from './emailEACP'
+import NoteAccessEACP from './noteAccessEACP'
+
+/**
+ * NoteOptions represents optional values that are not required for creating a note,
```
Comment (willmichael) [inline lib/types/eacp.js]: same copy pasta comment

File: lib/types/eacp.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+export default class EACP extends Serializable {
+  constructor(emailEACP, noteAccessEACP) {
+    super()
+
+    if (emailEACP instanceof EmailEACP) {
+      this.emailEACP = emailEACP
+    }
+    if (noteAccessEACP instanceof noteAccessEACP) {
+      this.noteAccessEACP = noteAccessEACP
+    }
```
Comment (willmichael) [inline lib/types/eacp.js]: this check fails silently if it's important down the line

File: lib/identity/index.js [hunk 8]
```diff
@@ -200,4 +248,68 @@ export default class Identity extends CryptoConsumer {
...
+      },
+      body: JSON.stringify(payload)
+    })
+    return validateResponseAsJSON(request)
+  }
+
+  async completeBrokerLogin(
+    authResponse,
+    noteId,
+    brokerUrl = `${this.config.apiUrl}/v1/identity/${this.config.realmId}/broker`
```
Comment (willmichael) [inline lib/identity/index.js]: can this be extracted into a class const?

File: lib/identity/index.js [hunk 9]
```diff
@@ -145,17 +145,65 @@ export default class Identity extends CryptoConsumer {
...
+        }
+      }
+    )
+    /* eslint-enable */
+
     return idClient
   }
 
-  async login(username, password) {
+  async login(username, password, broker = false) {
```
Comment (willmichael) [inline lib/identity/index.js]: Comment explaining that broker=true login is not meant for the typical user might be helpful here, or pulling it out into a separate function. It's not immediately clear why this broker bool is here and this is probably one of the most common functions a user of the sdk will calling.

File: lib/identity/index.js (hunk 2 above)
Comment (lkwdwrd) [reply lib/identity/index.js]: possibly -- but we don't fully understand the various use-cases right now to set up options correctly, so I think hard coded for our one use case is the way to go at this time, then we can make adjustments as it grows.

File: lib/identity/index.js (hunk 9 above)
Comment (lkwdwrd) [reply lib/identity/index.js]: Good point, all of these items need doc comments. I'll see what I can do about adding them.

File: lib/identity/index.js (hunk 8 above)
Comment (lkwdwrd) [reply lib/identity/index.js]: It relies on instance variables, but perhaps it could be extracted into a getter?

File: lib/types/eacp.js (hunk 6 above)
Comment (lkwdwrd) [reply lib/types/eacp.js]: yeah, looks like I need to do a code-comment pass. That's been killing me lately.

File: lib/types/eacp.js (hunk 7 above)
Comment (lkwdwrd) [reply lib/types/eacp.js]: Yeah -- I had it throw at one point. These are not required, though. I can do some fancy undefined checking, but went down that road for a bit and then turned back. This is they type of thing typescript will really help with.

File: lib/utils/credentials.js (hunk 3 above)
Comment (lkwdwrd) [reply lib/utils/credentials.js]: There are three notes:

- Login note
- Broker Key Note
- Broker Login Note

Each needs a different name, this helper gets the ID fro the two login notes, while the above 'key' note is the note containing the broker key. It gets a little muddy, so it's probably at least worth commenting to help clear it up a bit.

File: lib/identity/index.js (hunk 2 above)
Comment (galxy25) [reply lib/identity/index.js]: Agree, but would still prefer to see the broker flow be gated behind a Boolean parameter for this method that we can default to true, but thats a change we could just as easily make in the future 🤷🏻‍♀️

Comment (lkwdwrd) [reply lib/identity/index.js]: I started adding this, but I don't like any way it starts looking. I'm going to leave this hard-coded right now, and when we start doing our first alternate broker-based, or removing broker-based authentication we can refactor to support it.

File: lib/identity/index.js (hunk 8 above)
Comment (lkwdwrd) [reply lib/identity/index.js]: going to leave this as a direct param for now.

**Pull request merged**